### PR TITLE
feat: Read lambda layer property from terraform

### PR DESCRIPTION
**Pull request by mildaniel**
#### Why is this change necessary?
This feature follows the patterns we used to read and create lambda functions in the sam metadata file for lambda layers. It also updates the enrichment logic to enrich relevant layer properties and converts S3 code sources to their zip path counterparts.

#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [x] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [x] Write/update unit tests
- [ ] Write/update integration tests
- [ ] Write/update functional tests if needed
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: samcli/hook_packages/terraform/hooks/prepare.py [hunk 1]
```diff
@@ -2,7 +2,8 @@
 """
 Terraform prepare hook implementation
 """
-
+# pylint: disable=C0302
+# TODO: Move some of the logic out of this file and remove this disable
```
Comment (moelasmar) [inline samcli/hook_packages/terraform/hooks/prepare.py:6]: let's create task to do that

File: samcli/hook_packages/terraform/hooks/prepare.py [hunk 2]
```diff
@@ -304,7 +309,7 @@ def _validate_referenced_resource_matches_sam_metadata_type(
         )
 
 
-def _get_lambda_function_source_code_path(
+def _get_source_code_path(
```
Comment (moelasmar) [inline samcli/hook_packages/terraform/hooks/prepare.py:312]: can you update the comments, debug messages, and exception messages

File: samcli/hook_packages/terraform/hooks/prepare.py [hunk 3]
```diff
@@ -986,6 +1140,7 @@ def _get_s3_object_hash(bucket: str, key: str) -> str:
     md5 = hashlib.md5()
     md5.update(bucket.encode())
     md5.update(key.encode())
+    # TODO: Hash version if it exists in addition to key and bucket
```
Comment (moelasmar) [inline samcli/hook_packages/terraform/hooks/prepare.py:1143]: are you going to add it in a separate PR ?

File: samcli/hook_packages/terraform/hooks/prepare.py [hunk 4]
```diff
@@ -147,7 +150,9 @@ def _update_resources_paths(cfn_resources: Dict[str, Any], terraform_application
     terraform_application_dir: str
         The terraform application root directory where all paths will be relative to it
     """
-    resources_attributes_to_be_updated = {CFN_AWS_LAMBDA_FUNCTION: ["Code"]}
+    resources_attributes_to_be_updated = {
```
Comment (moelasmar) [inline samcli/hook_packages/terraform/hooks/prepare.py:153]: can you add a unit test case to cover updating the lambda layer paths in the mapped CFN resource dict. check this https://github.com/aws/aws-sam-cli/blob/8c8bc2e54ba1a03b5bc6d0a4cf7bb8d67ea8d3e3/tests/unit/hook_packages/terraform/test_prepare_hook.py#L2150

File: samcli/hook_packages/terraform/hooks/prepare.py [hunk 5]
```diff
@@ -569,6 +570,138 @@ def _enrich_image_lambda_function(
...
+    _set_zip_metadata_resources(
+        cfn_lambda_layer,
+        cfn_source_code_path,
+        output_directory_path,
+        terraform_application_dir,
+        CFN_CODE_PROPERTIES[CFN_AWS_LAMBDA_LAYER_VERSION],
+    )
+
+
+def _validate_referenced_resource_layer_matches_metadata_type(
```
Comment (moelasmar) [inline samcli/hook_packages/terraform/hooks/prepare.py:629]: I see that you unit test this function as part of the `_enrich_lambda_layer` .. I think it is better to have separate unit test cases for it, and also to unit test the exception scenarios

File: samcli/hook_packages/terraform/hooks/prepare.py [hunk 6]
```diff
@@ -594,6 +727,7 @@ def _enrich_resources_and_generate_makefile(
     resources_types_enrichment_functions = {
         "ZIP_LAMBDA_FUNCTION": _enrich_zip_lambda_function,
         "IMAGE_LAMBDA_FUNCTION": _enrich_image_lambda_function,
+        "LAMBDA_LAYER": _enrich_lambda_layer,
```
Comment (moelasmar) [inline samcli/hook_packages/terraform/hooks/prepare.py:730]: can you add a unit test case for testing enriching the lambda layers like this one https://github.com/aws/aws-sam-cli/blob/8c8bc2e54ba1a03b5bc6d0a4cf7bb8d67ea8d3e3/tests/unit/hook_packages/terraform/test_prepare_hook.py#L1360

File: samcli/hook_packages/terraform/hooks/prepare.py [hunk 7]
```diff
@@ -925,8 +1064,23 @@ def _check_image_config_value(image_config: Any) -> bool:
...
+}
+
 RESOURCE_TRANSLATOR_MAPPING: Dict[str, ResourceTranslator] = {
     TF_AWS_LAMBDA_FUNCTION: ResourceTranslator(CFN_AWS_LAMBDA_FUNCTION, AWS_LAMBDA_FUNCTION_PROPERTY_BUILDER_MAPPING),
+    TF_AWS_LAMBDA_LAYER_VERSION: ResourceTranslator(
+        CFN_AWS_LAMBDA_LAYER_VERSION, AWS_LAMBDA_LAYER_VERSION_PROPERTY_BUILDER_MAPPING
+    ),
+}
+
+CFN_CODE_PROPERTIES = {
```
Comment (moelasmar) [inline samcli/hook_packages/terraform/hooks/prepare.py:1081]: can we move these constant values to the beginning of the module.

File: samcli/hook_packages/terraform/hooks/prepare.py [hunk 8]
```diff
@@ -1002,8 +1157,11 @@ def _map_s3_sources_to_functions(s3_hash_to_source: Dict[str, str], cfn_resource
         CloudFormation resources
     """
     for _, resource in cfn_resources.items():
-        if resource.get("Type") == CFN_AWS_LAMBDA_FUNCTION:
-            code = resource.get("Properties").get("Code")
+        resource_type = resource.get("Type")
```
Comment (moelasmar) [inline samcli/hook_packages/terraform/hooks/prepare.py:1160]: can you add unit test cases for `_map_s3_source_to_functions` for the lambda layer use case. like this one https://github.com/aws/aws-sam-cli/blob/8c8bc2e54ba1a03b5bc6d0a4cf7bb8d67ea8d3e3/tests/unit/hook_packages/terraform/test_prepare_hook.py#L882

**Review by moelasmar — CHANGES_REQUESTED** [review]
it looks good to me, just some minor comments, and adding some missing unit test cases.

File: samcli/hook_packages/terraform/hooks/prepare.py (hunk 3 above)
Comment (mildaniel) [reply samcli/hook_packages/terraform/hooks/prepare.py:1143]: Yes I'll raise a separate PR after this one to keep the size of this one smaller. I also created an internal ticket to track this item.

File: samcli/hook_packages/terraform/hooks/prepare.py (hunk 1 above)
Comment (mildaniel) [reply samcli/hook_packages/terraform/hooks/prepare.py:6]: Done.

File: samcli/hook_packages/terraform/hooks/prepare.py (hunk 4 above)
Comment (mildaniel) [reply samcli/hook_packages/terraform/hooks/prepare.py:153]: I added some layers to this test since it was already generic enough.

**Review by moelasmar — APPROVED** [review]

**Review by lucashuy — APPROVED** [review]

**Pull request merged**